### DISCUSSION TITLE
Get zoom level

### DISCUSSION THREAD

**Discussion by kboshold** [wezterm/wezterm#3959] [Q&A] [answered]
Is it possible to get the current zoom level using a shell script. I want to show the current zoom in my TMUX statusbar.

**Comment by wez** [comment]
Do you mean the font size scaling factor or something else?
The font size is a property of the gui and isn't accessible via the CLI, which operates on the mux layer.

**Comment by nekowinston** [comment, marked as answer]
You *could* keep track of it by overriding the default Zoom action, and storing data in `wezterm.GLOBAL`:

```lua
local wezterm = require("wezterm")
local act = wezterm.action
local c = wezterm.config_builder()

if wezterm.GLOBAL.zoom_factor == nil then
  wezterm.GLOBAL.zoom_factor = 1.0
end

local increaseFontSize = wezterm.action_callback(function(window, pane)
  wezterm.GLOBAL.zoom_factor = wezterm.GLOBAL.zoom_factor + 0.1
  wezterm.log_info("zoom_factor: " .. wezterm.GLOBAL.zoom_factor)
  window:perform_action(act.IncreaseFontSize, pane)
end)

local decreaseFontSize = wezterm.action_callback(function(window, pane)
  wezterm.GLOBAL.zoom_factor = wezterm.GLOBAL.zoom_factor - 0.1
  wezterm.log_info("zoom_factor: " .. wezterm.GLOBAL.zoom_factor)
  window:perform_action(act.DecreaseFontSize, pane)
end)

local resetFontSize = wezterm.action_callback(function(window, pane)
  wezterm.GLOBAL.zoom_factor = 1.0
  wezterm.log_info("zoom_factor: " .. wezterm.GLOBAL.zoom_factor)
  window:perform_action(act.ResetFontSize, pane)
end)

wezterm.on("update-right-status", function(window, pane)
  window:set_right_status("Zoom " .. tostring(wezterm.GLOBAL.zoom_factor))
end)

c.keys = {
  { key = "=", mods = "CTRL", action = increaseFontSize },
  { key = "=", mods = "SHIFT|CTRL", action = increaseFontSize },
  { key = "=", mods = "SUPER", action = increaseFontSize },
  { key = "-", mods = "CTRL", action = decreaseFontSize },
  { key = "-", mods = "SHIFT|CTRL", action = decreaseFontSize },
  { key = "-", mods = "SUPER", action = decreaseFontSize },
  { key = "0", mods = "CTRL", action = resetFontSize },
  { key = "0", mods = "SHIFT|CTRL", action = resetFontSize },
  { key = "0", mods = "SUPER", action = resetFontSize },
}

return c
```

~~I should note that this is a bit unreliable, since it resets when going to the debug overlay; but as long as you're navigating across panes and tabs, it doesn't seem to be too far off for me.~~ (See wez's comment below for why this occured!)

**Reply by kboshold** [reply]
@nekowinston wow thanks for your answer, that I will try directly!.. How can I access `wezterm.GLOBAL` again later (outside the wezterm / sh script) context? 

Otherwise I can of course access the log here or write it to a separate file if necessary (like `.cache/wezterm/zoom_level`).

**Reply by nekowinston** [reply]
As my edit says, the `wezterm.log_info` here actually isn't reliable sadly, but you can access `GLOBAL` in your config like I'm doing here with the `"update-right-status"` event. See more about `GLOBAL` [here](https://wezfurlong.org/wezterm/config/lua/wezterm/GLOBAL.html).

As another example, I use `GLOBAL` to add a custom keybind that toggles the tab bar in [my config](https://github.com/nekowinston/dotfiles/blob/a8b40fadac8aeb21da97fa3847c246681546e22d/home/apps/wezterm/keys.lua#L17-L23).

**Reply by wez** [reply]
I want to note about this unconditional assignment at the file scope at the top of the file:

```lua
wezterm.GLOBAL.zoom_factor = 1.0
```

every time any new lua context is created, which can be multiple times during startup, config file changes, window creation, event handling and potentially any time a background thread needs to reason about the config, those file-scoped statements will be evaluated, and that will cause your zoom_factor to be reset to 1.0 when that happens.

When you are using normal lua global variables, this doesn't matter so much, as they are no shared between lua contexts, but since `wezterm.GLOBAL` is explicitly globalized, the unconditional assignment is likely undesirable.

I would suggest that you avoid assigning that way, and either:

## Assume a default when you read the value

Better when you have fewer places that depend upon it, but here's an example:

```lua
local increaseFontSize = wezterm.action_callback(function(window, pane)
  wezterm.GLOBAL.zoom_factor = (wezterm.GLOBAL.zoom_factor or 1.0) + 0.1
  wezterm.log_info("zoom_factor: " .. wezterm.GLOBAL.zoom_factor)
  window:perform_action(act.IncreaseFontSize, pane)
end)
```

(This is the technique used in the example on https://wezfurlong.org/wezterm/config/lua/wezterm/GLOBAL.html)

## Use a conditional initial assignment

Change the unconditional assignment so that it only sets the default when it isn't set:

```lua
if wezterm.GLOBAL.zoom_factor == nil then
  wezterm.GLOBAL.zoom_factor = 1.0
end
```

# Another thing!

The zoom level is a per-gui-window property, so if you have multiple windows the above will misbehave.
You can approximately resolve that by using `window:window_id()` and something like:

```lua
if wezterm.GLOBAL.zoom_factor == nil then
  wezterm.GLOBAL.zoom_factor = {}
end

local increaseFontSize = wezterm.action_callback(function(window, pane)
   local id = window:window_id()
   wezterm.GLOBAL.zoom_factor[id] = (wezterm.GLOBAL.zoom_factor[id] or 1.0) + 0.1
   -- ..
end
```

but note that `window:window_id()` returns the underlying multiplexer window ID rather than a GUI id (wezterm doesn't currently have gui window ids), and the multiplexer id can change when the active workspace is changed!

**Reply by nekowinston** [reply]
Thanks for the additional context!
I didn't know that opening the REPL created a new lua context (which makes sense of course), and I forgot about conditionally assigning here. I updated my original comment.
